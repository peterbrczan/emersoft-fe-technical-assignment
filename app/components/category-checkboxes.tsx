'use client';

import type { FC } from 'react';
import React, { useEffect, useState } from 'react';
import { Category } from '@/app/models/types/category';
import { useTranslations } from 'use-intl';

export type CategoryCheckboxesProps = {
  categories: Category[];
  handleCategorySlugs: (categorySlugs: string[]) => void;
};

export const CategoryCheckboxes: FC<CategoryCheckboxesProps> = (props) => {
  const [categorySlugs, setCategorySlugs] = useState<string[]>([]);
  const textCategory = useTranslations('Category');

  useEffect(() => {
    props.handleCategorySlugs(categorySlugs);
  }, [categorySlugs]);

  function toggleCheckbox(categorySlug: string) {
    let copyCategorySlugs = [...categorySlugs];

    if (categorySlugs.includes(categorySlug)) {
      copyCategorySlugs = copyCategorySlugs.filter((copyCategorySlug) => copyCategorySlug !== categorySlug);
    } else {
      copyCategorySlugs.push(categorySlug);
    }

    setCategorySlugs(copyCategorySlugs);
  }

  return (
    <div className="flex gap-4 flex-wrap">
      {props.categories.map((category) => (
        <div key={category.slug} className="flex items-center">
          <input
            id={category.slug}
            type="checkbox"
            checked={categorySlugs.includes(category.slug)}
            className="w-[16px] h-[16px] rounded-[4px] cursor-pointer"
            onChange={() => toggleCheckbox(category.slug)}
          />
          <label for={category.slug} className="ml-[8px] text-sm font-medium text-gray-900 cursor-pointer">
            {textCategory(category.name)}
          </label>
        </div>
      ))}
    </div>
  );
};
