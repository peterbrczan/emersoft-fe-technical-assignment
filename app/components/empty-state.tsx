import type { FC, PropsWithChildren } from 'react';
import React from 'react';

export const EmptyState: FC<PropsWithChildren> = ({ children }) => {
  return <div className="bg-gray-200 h-32 p-2 flex items-center justify-center rounded-[4px]">{children}</div>;
};
