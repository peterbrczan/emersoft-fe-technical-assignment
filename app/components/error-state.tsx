import type { FC, PropsWithChildren } from 'react';
import React from 'react';

export const ErrorState: FC<PropsWithChildren> = ({ children }) => {
  return <div className="bg-red-300 h-32 p-2 flex items-center justify-center rounded-[4px]">{children}</div>;
};
