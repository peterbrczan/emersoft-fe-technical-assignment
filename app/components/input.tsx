'use client';

import type { FC } from 'react';
import React, { useEffect, useState } from 'react';
import { useTranslations } from 'use-intl';
import { useDebounce } from 'use-debounce';

export type InputProps = {
  handleSearchValue: (value: string) => void;
};

export const Input: FC<InputProps> = (props) => {
  const [value, setValue] = useState('');
  const [searchValue] = useDebounce(value, 500);

  useEffect(() => {
    props.handleSearchValue(searchValue);
  }, [searchValue]);

  const textCommon = useTranslations('Common');

  return (
    <div className="relative">
      <div className="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
        <svg className="w-4 h-4 text-gray-500" fill="none" viewBox="0 0 20 20">
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
          />
        </svg>
      </div>
      <input
        type="search"
        className="block w-full p-2 pl-8 text-sm text-gray-900 border border-gray-300 rounded-[4px] bg-gray-100 "
        placeholder={textCommon('Search by a title')}
        onChange={({ currentTarget: { value } }) => setValue(value)}
        data-testid="search-input"
      />
    </div>
  );
};
