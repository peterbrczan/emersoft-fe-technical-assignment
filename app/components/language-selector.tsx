'use client';

import type { FC } from 'react';
import React, { useMemo } from 'react';
import { useLocale } from 'use-intl';
import { Locale } from '@/app/models/enums/locale';
import Link from 'next/link';
import { usePathname } from 'next/navigation';

export const LanguageSelector: FC = () => {
  const locale = useLocale() as Locale;
  const pathname = usePathname();

  const flag = useMemo(() => {
    switch (locale) {
      case Locale.HU: {
        return { src: 'https://flagcdn.com/w40/hu.png', alt: 'Hungarian Flag' };
      }
      case Locale.EN:
      default: {
        return {
          src: 'https://flagcdn.com/w40/gb.png',
          alt: 'Great Britain flag',
        };
      }
    }
  }, [locale]);

  function getNextHref() {
    const nextUrl = pathname;
    if (nextUrl.startsWith(`/${Locale.HU}`)) {
      return nextUrl.replace(`/${Locale.HU}`, `/${Locale.EN}`);
    } else if (nextUrl.startsWith(`/${Locale.EN}`)) {
      return nextUrl.replace(`/${Locale.EN}`, `/${Locale.HU}`);
    } else {
      return `${Locale.HU}${nextUrl}`;
    }
  }

  return (
    <Link href={getNextHref()} className="w-[20px] h-[20px] rounded-full">
      <img className="w-[20px] h-[20px] rounded-full" src={flag.src} alt={flag.alt} />
    </Link>
  );
};
