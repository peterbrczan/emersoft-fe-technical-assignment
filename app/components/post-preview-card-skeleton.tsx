import type { FC } from 'react';
import React from 'react';
import { Skeleton } from '@/app/components/skeleton';

export const PostPreviewCardSkeleton: FC = () => {
  return (
    <div className="max-w-xs w-[100%] h-[100%] bg-white shadow-md rounded-[4px] overflow-hidden ">
      <Skeleton className="w-full h-48" />
      <div className="px-2 py-2">
        <div className="flex flex-wrap gap-1 mb-2">
          {new Array(2).fill('').map((_, index) => (
            <Skeleton key={index} className="h-[28px] w-[100px]" />
          ))}
        </div>
        <div className="mb-2">
          <Skeleton className="h-[28px] w-[200px] mb-1" />
          <Skeleton className="h-[28px] w-[150px]" />
        </div>
      </div>
    </div>
  );
};
