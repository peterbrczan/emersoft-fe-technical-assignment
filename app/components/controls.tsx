import type { FC } from 'react';
import React, { useEffect, useState } from 'react';
import { Input } from '@/app/components/input';
import { useGetBlog } from '@/app/hooks/use-get-blog';
import { Pagination } from '@/app/components/pagination';
import { CategoryCheckboxes } from '@/app/components/category-checkboxes';
import { Category } from '@/app/models/types/category';

export const Controls: FC = (props) => {
  const [searchValue, setSearchValue] = useState('');
  const [pageNumber, setPageNumber] = useState(1);
  const [categorySlugs, setCategorySlugs] = useState<string[]>([]);
  const [categories, setCategories] = useState<Category[]>([]);

  const getBlog = useGetBlog({ searchValue, pageNumber, categorySlugs });

  useEffect(() => {
    if (getBlog.isFetched) {
      setCategories(getBlog.data?.blog.categories ?? []);
    }
  }, [getBlog.data, getBlog.isFetched]);

  function handleSearchValue(searchValue: string) {
    // Resetting the page number
    setPageNumber(1);
    setSearchValue(searchValue);
  }

  function handleCategorySlugs(categorySlugs: string[]) {
    // Resetting the page number
    setPageNumber(1);
    setCategorySlugs(categorySlugs);
  }

  return (
    <div className="grid grid-cols-[1fr_auto] gap-4 items-center justify-between">
      <Input handleSearchValue={handleSearchValue} />
      <Pagination
        pageNumber={pageNumber}
        numberOfTotalPages={getBlog.data?.numberOfTotalPages ?? 0}
        handlePageNumber={setPageNumber}
      />
      <CategoryCheckboxes categories={categories} handleCategorySlugs={handleCategorySlugs} />
    </div>
  );
};
