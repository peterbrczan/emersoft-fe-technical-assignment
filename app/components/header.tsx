'use client';

import type { FC } from 'react';
import React from 'react';
import { useTranslations } from 'use-intl';
import { LanguageSelector } from '@/app/components/language-selector';

export const Header: FC = () => {
  const textCommon = useTranslations('Common');

  return (
    <div className="border-b-2 border-gray-100 min-h-[80px] grid grid-cols-[1fr_auto] justify-between p-4 gap-[4px]">
      <div className="self-center flex flex-wrap gap-2 items-center">
        <h1 className="text-2xl font-extrabold">Emersoft</h1>
        <div className="w-[8px] h-[8px] rounded-full bg-pink-700"></div>
        <p>{textCommon('Frontend Technical Assignment')}</p>
      </div>
      <div className="justify-self-end self-center">
        <LanguageSelector />
      </div>
    </div>
  );
};
