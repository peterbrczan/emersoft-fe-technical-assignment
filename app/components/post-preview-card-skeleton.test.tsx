import { render } from '@/app/utils/test-renderer';
import { PostPreviewCardSkeleton } from '@/app/components/post-preview-card-skeleton';

describe('[component]: post-preview-card-skeleton', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<PostPreviewCardSkeleton />);

    expect(asFragment()).toMatchSnapshot();
  });
});
