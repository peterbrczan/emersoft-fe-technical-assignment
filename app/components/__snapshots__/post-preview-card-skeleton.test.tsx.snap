// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[component]: post-preview-card-skeleton should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-xs w-[100%] h-[100%] bg-white shadow-md rounded-[4px] overflow-hidden "
  >
    <div
      class="animate-pulse  bg-gray-100 rounded-md dark:bg-gray-300 w-full h-48"
    />
    <div
      class="px-2 py-2"
    >
      <div
        class="flex flex-wrap gap-1 mb-2"
      >
        <div
          class="animate-pulse  bg-gray-100 rounded-md dark:bg-gray-300 h-[28px] w-[100px]"
        />
        <div
          class="animate-pulse  bg-gray-100 rounded-md dark:bg-gray-300 h-[28px] w-[100px]"
        />
      </div>
      <div
        class="mb-2"
      >
        <div
          class="animate-pulse  bg-gray-100 rounded-md dark:bg-gray-300 h-[28px] w-[200px] mb-1"
        />
        <div
          class="animate-pulse  bg-gray-100 rounded-md dark:bg-gray-300 h-[28px] w-[150px]"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
